File: dnsApi/apikey/perl/dns_lookup_apikey.pl
use strict;
use warnings;
use Data::Dumper;
use LWP::Simple;                            # From CPAN
use JSON qw( decode_json encode_json );     # From CPAN
use Time::HiRes qw( time );                 # From CPAN
use Digest::HMAC_MD5 qw( hmac_md5_hex );    # From CPAN
use URI::Escape;                            # From CPAN
use MIME::Base64 qw( encode_base64 );

my @domains = (
    'google.com',
    'example.com',
    'whoisxmlapi.com',
    'twitter.com'
);
my $url = 'https://whoisxmlapi.com/whoisserver/DNSService?';
my $username = 'username';
my $apiKey = 'api_key';
my $secret = 'secret_key';

my $timestamp = int((time * 1000 + 0.5));
my $digest = generateDigest($username, $timestamp, $apiKey, $secret);

foreach my $domain (@domains){
    my $requstString = buildRequest(
        $username,
        $timestamp,
        $digest,
        $domain
    );
    my $response = get($url . $requstString);
    if (index($response, 'Request timeout')){
        $timestamp = int((time * 1000 + 0.5));
        $digest = generateDigest($username, $timestamp, $apiKey, $secret);
        $requstString = buildRequest(
            $username,
            $timestamp,
            $digest,
            $domain
        );
        $response = get($url . $requstString);
    }
    printResponse($response);
}

sub generateDigest{
    my ($username, $timestamp, $apiKey, $secret) = @_;

    my $digest = $username . $timestamp . $apiKey;
    my $hash = hmac_md5_hex($digest, $secret);
    return uri_escape($hash);
}

sub buildRequest{
    my ($username, $timestamp, $digest, $domain) = @_;
    my $requestString = 'requestObject=';
    my %request =(
        'u' => $username,
        't' => $timestamp
    );
    my $requestJson = encode_json(\%request);
    my $requestBase64 = encode_base64($requestJson);
    $requestString .= uri_escape($requestBase64);
    $requestString .= '&type=_all';
    $requestString .= '&digest=';
    $requestString .= $digest;
    $requestString .= '&domainName='
        . $domain
        . '&outputFormat=json';
    return $requestString;
}

sub printResponse{
    my ($response) = @_;
    print $response;
    print "---------------------------------------\n";
}

File: dnsApi/apikey/nodejs/dns_lookup_apikey.js
const https = require('https');
const queryString = require('querystring');
const crypto = require('crypto');

const url = 'https://whoisxmlapi.com/whoisserver/DNSService?';
const username = 'username';
const apiKey = 'api_key';
const secretKey = 'secret_key';

const domains = [
    'google.com',
    'example.com',
    'whoisxmlapi.com',
    'twitter.com'
];

for(var i in domains) {
    getWhois(username, apiKey, secretKey, domains[i]);
}

function getWhois(username, apiKey, secretKey, domain)
{
    timestamp = (new Date).getTime();
    digest = generateDigest(username, timestamp, apiKey, secretKey);
    var requestString = buildRequest(username, timestamp, digest, domain);
    https.get(url + requestString, function (res) {
        const statusCode = res.statusCode;

        if (statusCode !== 200) {
            console.log('Request failed: '
                + statusCode
            );
        }

        var rawData = '';

        res.on('data', function(chunk) {
            rawData += chunk;
        });

        res.on('end', function () {
            console.log(rawData);
        })
    }).on('error', function(e) {
        console.log("Error: " + e.message);
    });

}

function generateDigest(username, timestamp, apiKey, secretKey) {
    var data = username + timestamp + apiKey;
    var hmac = crypto.createHmac('md5', secretKey);
    hmac.update(data);
    return hmac.digest('hex');
}

function buildRequest(username, timestamp, digest, domain) {
    var data = {
        u: username,
        t: timestamp
    };

    var dataJson = JSON.stringify(data);
    var dataBase64 = Buffer.from(dataJson).toString('base64');

    var request = {
        requestObject: dataBase64,
        type: '_all',
        digest: digest,
        domainName: domain,
        outputFormat: 'json'
    };

    return queryString.stringify(request);
}